pjsip/src/pjsua-lib/pjsua_pres.c

```diff
@@ -349,128 +349,142 @@ pjsua_buddy_get_dlg_event_info( pjsua_buddy_id buddy_id,
-                   buddy->dlg_ev_status.info[0].dialog_state.slen);
-        total += info->dialog_state.slen;
-
-        info->dialog_duration.ptr = info->buf_ + total;
-        pj_strncpy(&info->dialog_duration,
-                   &buddy->dlg_ev_status.info[0].dialog_duration,
-                   buddy->dlg_ev_status.info[0].dialog_duration.slen);
-        total += info->dialog_duration.slen;
-
-        if (buddy->dlg_ev_status.info[0].local_identity.ptr) {
+        if (total < sizeof(info->buf_)) {
+            info->dialog_id.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_id,
+                       &buddy->dlg_ev_status.info[0].dialog_id,
+                       sizeof(info->buf_)-total);
+            total += info->dialog_id.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_call_id.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_call_id,
+                       &buddy->dlg_ev_status.info[0].dialog_call_id,
+                       sizeof(info->buf_)-total);
+            total += info->dialog_call_id.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_remote_tag.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_remote_tag,
+                       &buddy->dlg_ev_status.info[0].dialog_remote_tag,
+                       sizeof(info->buf_)-total);
+            total += info->dialog_remote_tag.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_local_tag.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_local_tag,
+                       &buddy->dlg_ev_status.info[0].dialog_local_tag,
+                       sizeof(info->buf_)-total);
+            total += info->dialog_local_tag.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_direction.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_direction,
+                       &buddy->dlg_ev_status.info[0].dialog_direction,
+                       sizeof(info->buf_)-total);
+            total += info->dialog_direction.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_state.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_state,
+                       &buddy->dlg_ev_status.info[0].dialog_state,
+                       buddy->dlg_ev_status.info[0].dialog_state.slen);
+            total += info->dialog_state.slen;
+        }
+
+        if (total < sizeof(info->buf_)) {
+            info->dialog_duration.ptr = info->buf_ + total;
+            pj_strncpy(&info->dialog_duration,
+                       &buddy->dlg_ev_status.info[0].dialog_duration,
+                       buddy->dlg_ev_status.info[0].dialog_duration.slen);
+            total += info->dialog_duration.slen;
+        }
+
+        if (buddy->dlg_ev_status.info[0].local_identity.slen > 0 &&
+            total < sizeof(info->buf_))
+        {
```

Comment:
- Raising assertion and/or printing log when buffer is insufficient may be useful for developer?
- The checks & string copy ops seem similar, perhaps a helper function or a macro can improve readibility & correctness, something like:
```
#define COPY_BUF(field)
  if (status->field.slen > 0) {
    if (total + status->field.slen < size(info->buf_)) {
      info->field.ptr = info->buf_ + total;
      strncpy(..);
      total += field.slen;
    } else {
      PJ_LOG(..);
      pj_assert(..);
    }
  }
```

Reply:
Great idea!
Now it becomes much more compact.

I don't think there should be an assertion there though, since if remote sends something that's larger than the buffer size, we don't want to assert.